service/worker/scanner/build_ids/scavenger.go

```diff
@@ -282,12 +293,16 @@ func (a *Activities) findBuildIdsToRemove(
 			}
 			buildIdIsSetDefault := buildIdIdx == len(set.BuildIds)-1
 			setIsQueueDefault := setIdx == len(versioningData.VersionSets)-1
-			// Don't remove if build id is the queue default of there's another active build id in this set.
+			// Don't remove if build id is the queue default or there's another active build id in
+			// this set, since we might need to dispatch new tasks to this set. But if no build ids
+			// are active for the whole set, we can remove them all.
 			if buildIdIsSetDefault && (setIsQueueDefault || setActive > 1) {
 				continue
 			}
-			timeSinceWasDefault := time.Since(hybrid_logical_clock.UTC(*buildId.BecameDefaultTimestamp))
-			if timeSinceWasDefault < a.removableBuildIdDurationSinceDefault() {
+			if hlc.SincePtr(buildId.BecameDefaultTimestamp) < removableBuildIdDurationSinceDefault {
+				continue
+			}
+			if hlc.SincePtr(buildId.StateUpdateTimestamp) < retention {
```

Comment:
another thought here: suppose we add a STATE_BAD: should we wait a full retention after the state change from ACTIVE to BAD before we consider it for cleanup? or do we need yet another timestamp? I'm thinking just wait the retention period from marking BAD.

(or do we only update StateUpdateTimestamp on transition to DELETED, not BAD?)

Reply:
The transition to a bad state is going to be so rare that I think we can either always check BAD versions or wait from the time of the state update.